languages/csharp/exercises/concept/randomness/RandomnessTests.cs

```diff
@@ -0,0 +1,21 @@
+using Xunit;
+using System;
+
+public class UserDefinedExceptionsTests
+{
+    [Fact]
+    public void RollDie()
+    {
+        var player = new Player();
+        var pips = player.RollDie();
```

Comment:
One thing we could do here is to call `RollDie` a lot of times (e.g. 100 times or so), to verify that they all are in the right range. Another test that we could add is to call `RollDie` a lot of times and verify that it won't return the same value each time (to make sure students are not using a seed).

Reply:
>  verify that they all are in the right range.

I assumed that we would have a few representations which would take care of this.  I am loth to add a test that will occasionally give a false positive (or negative or whichever way it is).  

Are you thinking that we would just make multiple calls in a test?   If we include a hash set in the test file I would like to avoid adding the hash set as a dependency.  As I think I said before I did half the haskell course without understanding the test framework and it really didn't matter.  

>  verify that it won't return the same value each time (to make sure students are not using a seed).

Similar comment I suppose.  I was toying with the idea of getting the student to subclass `Random` and generate `enum`s and return that to the caller, but it doesn't get us much further than creating two players (assuming they are not static, and rolling the die 3 or 4 times.

Reply:
> I am loth to add a test that will occasionally give a false positive (or negative or whichever way it is).

This could indeed be the case for the uniqueness test, so let's not include that. However, the "is the value in the right range" test should not be prone to false positives (or negatives), or am I missing something?

Reply:
>  am I missing something?

The value we're really interested in is 18 (one below the upper bound) and possibly ensuring that 0 does not turn up.  Representations are going to give us our reassurance there as we can't rely on a test showing this.  Have I missed the point about representations?

Reply:
> The value we're really interested in is 18 (one below the upper bound) and possibly ensuring that 0 does not turn up. Representations are going to give us our reassurance there as we can't rely on a test showing this. Have I missed the point about representations?

I get the feeling we're arguing different things :) This test verifies that the output of the `RollDie()` method is in the range 1-18. This is good. As we're dealing with randomness, it could be that the student has set an invalid upper bound. Suppose that upper bound is 19, that would mean that this test fails only 1/19th of a time. Calling the `RollDie()` method 100 times greatly reduces this chance (although it doesn't eliminate it).

> Representations are going to give us our reassurance there as we can't rely on a test showing this

Well yes, but in general anything that we _can_ do to help guide students towards the right solution without it involving an analyzer or representation is a win in my mind.

Reply:
> it could be that the student has set an invalid upper bound. Suppose that upper bound is 19, that would mean that this test fails only 1/19th of a time

We loop 100 times.  The problem is that it does not catch the most likely error.  The upper bound is exclusive so the most common mistake is to set it to 18 (inclusive) rather than 19 which is what it should be.  We cannot easily detect that 18 never turns up. 

It does catch the second most likely error: setting the lower bound to 0 either explicitly or implicitly.

> but in general anything that we can do to help guide students towards the right solution without it involving an analyzer or representation is a win in my mind.

I agree with the sentiment but I would understand  this better if I knew the workflow.  If a student is operating in the browser do they run all the tests  by pressing a button or something after they have completed all tasks?